File: device/cuda/quant/native/quant.c
#include "quant_dispatch.h"
#include "../internal/bridge/core_private.h"

#include <stdio.h>

static const char* g_cuda_quant_module_source = NULL;

void cuda_quant_register_module_source(const char* source) {
    g_cuda_quant_module_source = source;
}

const char* cuda_quant_module_source(void) {
    return g_cuda_quant_module_source;
}

int cuda_dispatch_quantization(
    CUDADeviceRef contextRef,
    int elementDType,
    CUDABufferRef inputRef,
    CUDABufferRef outRef,
    float invScale,
    int zeroPoint,
    uint32_t count,
    uint64_t completionToken,
    CUDAStatus* status
) {
    cuda_status_clear(status);

    if (count == 0) {
        return 0;
    }

    if (inputRef == NULL || outRef == NULL) {
        cuda_status_set(status, -2, "nil CUDA buffer");
        return -2;
    }

    const char* suffix = cuda_element_dtype_suffix(elementDType);

    if (suffix == NULL) {
        cuda_status_set(status, -6, "unknown CUDA quant dtype");
        return -6;
    }

    char kernelName[128];
    int nameCode = cuda_compose_kernel_name(kernelName, sizeof(kernelName), "int8_quant", suffix, status);

    if (nameCode != 0) {
        return nameCode;
    }

    const char* moduleSource = cuda_quant_module_source();

    if (moduleSource == NULL) {
        cuda_status_set(status, -7, "CUDA quant module source not registered");
        return -7;
    }

    CUDAContext* context = NULL;
    CUDAStreamRef stream = NULL;
    int prepareCode = cuda_context_prepare(contextRef, status, &context, &stream);

    if (prepareCode != 0) {
        return prepareCode;
    }

    CUDAKernelRef kernel = cuda_get_kernel(context, moduleSource, kernelName, status);

    if (kernel == NULL) {
        return status != NULL && status->code != 0 ? status->code : -7;
    }

    void* inputPtr = cuda_buffer_device_ptr(inputRef);
    void* outPtr = cuda_buffer_device_ptr(outRef);
    void* args[] = {&inputPtr, &outPtr, &invScale, &zeroPoint, &count};
    uint32_t launchCount = cuda_vector_launch_count(count, elementDType);
    int launchCode = cuda_launch_1d(context, kernel, stream, launchCount, args, sizeof(args), status);

    if (launchCode != 0) {
        return launchCode;
    }

    cuda_track_completion(context, stream, completionToken, NULL, status);
    return 0;
}

File: device/metal/quant/native/quant.m
#include "quant.h"

#include "../internal/bridge/core_private.h"

#include <Foundation/Foundation.h>
#include <Metal/Metal.h>
#include <stdio.h>

static void metal_quantization_status_clear(MetalStatus* status) {
    if (status == NULL) {
        return;
    }

    status->code = 0;
    status->message[0] = '\0';
}

static void metal_quantization_status_set(MetalStatus* status, int code, const char* message) {
    if (status == NULL) {
        return;
    }

    status->code = code;

    if (message == NULL) {
        status->message[0] = '\0';
        return;
    }

    snprintf(status->message, METAL_STATUS_MESSAGE_BYTES, "%s", message);
}

static const char* metal_quantization_kernel_name(int operation) {
    switch (operation) {
    case 0: return "int8_dequant";
    case 1: return "int4_dequant";
    case 2: return "int8_quant";
    default: return NULL;
    }
}


